File: lib/providers/hubspot.ts
/**
 * HubSpot CRM provider — upsert a contact + attach a note from an
 * enriched prospect / completed campaign. Read by the `push_to_crm`
 * agent tool.
 *
 * Auth: a HubSpot Private App token in `HUBSPOT_API_KEY`. Without it
 * we mock cleanly so the whole flow runs key-free (matching every
 * other provider).
 *
 * API reference:
 *   POST /crm/v3/objects/contacts            (create)
 *   POST /crm/v3/objects/contacts/{id}       (update by id)
 *   GET  /crm/v3/objects/contacts/{email}?idProperty=email
 *   POST /crm/v3/objects/notes               (create note + association)
 */

import {
  clampNoteBody,
  contactToHubSpotProperties,
  hubspotConfigured,
  isValidEmail,
  mockContactId,
  mockNoteId,
  type CrmContactInput,
  type CrmNoteInput,
} from "@/lib/providers/hubspot-core"

const API_BASE = "https://api.hubapi.com"
const REQUEST_TIMEOUT_MS = 12_000

export interface PushContactResult {
  ok: boolean
  contact_id: string
  created: boolean
  mock: boolean
  error?: string
}

export interface AddNoteResult {
  ok: boolean
  note_id: string
  mock: boolean
  error?: string
}

export function isHubSpotConfigured(): boolean {
  return hubspotConfigured(process.env.HUBSPOT_API_KEY)
}

function authHeader(): Record<string, string> {
  return {
    Authorization: `Bearer ${process.env.HUBSPOT_API_KEY}`,
    "Content-Type": "application/json",
  }
}

async function hubFetch(
  path: string,
  init: RequestInit,
): Promise<{ ok: boolean; status: number; json: unknown }> {
  try {
    const res = await fetch(`${API_BASE}${path}`, {
      ...init,
      headers: { ...authHeader(), ...(init.headers ?? {}) },
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    })
    const json = await res.json().catch(() => ({}))
    return { ok: res.ok, status: res.status, json }
  } catch (err) {
    return { ok: false, status: 0, json: { message: (err as Error).message } }
  }
}

/**
 * Upsert a contact by email. HubSpot's create endpoint returns 409 on
 * duplicate; we fall through to a GET by `idProperty=email` to fetch
 * the existing id and then PATCH. Three calls in the duplicate path,
 * one in the happy path — well within the free-tier rate limit.
 */
export async function pushContact(input: CrmContactInput): Promise<PushContactResult> {
  if (!isValidEmail(input.email)) {
    return {
      ok: false,
      contact_id: "",
      created: false,
      mock: false,
      error: "invalid email",
    }
  }

  if (!isHubSpotConfigured()) {
    return {
      ok: true,
      contact_id: mockContactId(input.email),
      created: true,
      mock: true,
    }
  }

  const properties = contactToHubSpotProperties(input)

  const create = await hubFetch("/crm/v3/objects/contacts", {
    method: "POST",
    body: JSON.stringify({ properties }),
  })
  if (create.ok) {
    const id = readId(create.json)
    return { ok: true, contact_id: id ?? "", created: true, mock: false }
  }
  if (create.status === 409) {
    const existing = await hubFetch(
      `/crm/v3/objects/contacts/${encodeURIComponent(properties.email)}?idProperty=email`,
      { method: "GET" },
    )
    const id = readId(existing.json)
    if (!id) {
      return {
        ok: false,
        contact_id: "",
        created: false,
        mock: false,
        error: "duplicate but lookup failed",
      }
    }
    const update = await hubFetch(`/crm/v3/objects/contacts/${id}`, {
      method: "PATCH",
      body: JSON.stringify({ properties }),
    })
    return {
      ok: update.ok,
      contact_id: id,
      created: false,
      mock: false,
      error: update.ok ? undefined : `HubSpot ${update.status}`,
    }
  }
  return {
    ok: false,
    contact_id: "",
    created: false,
    mock: false,
    error: `HubSpot ${create.status}: ${readMessage(create.json)}`,
  }
}

export async function addNote(
  contactId: string,
  note: CrmNoteInput,
): Promise<AddNoteResult> {
  if (!contactId) {
    return { ok: false, note_id: "", mock: false, error: "missing contact id" }
  }

  if (!isHubSpotConfigured()) {
    return {
      ok: true,
      note_id: mockNoteId(`${contactId}:${note.body}`),
      mock: true,
    }
  }

  const ts = note.timestamp ?? new Date().toISOString()
  const body = clampNoteBody(note.body)

  // HubSpot's "note" engagement requires a hs_note_body property + a
  // contact-to-note association (assoc type 202).
  const res = await hubFetch("/crm/v3/objects/notes", {
    method: "POST",
    body: JSON.stringify({
      properties: {
        hs_note_body: body,
        hs_timestamp: ts,
      },
      associations: [
        {
          to: { id: contactId },
          types: [
            {
              associationCategory: "HUBSPOT_DEFINED",
              associationTypeId: 202,
            },
          ],
        },
      ],
    }),
  })
  const id = readId(res.json)
  return {
    ok: res.ok && Boolean(id),
    note_id: id ?? "",
    mock: false,
    error: res.ok ? undefined : `HubSpot ${res.status}: ${readMessage(res.json)}`,
  }
}

function readId(payload: unknown): string | null {
  if (!payload || typeof payload !== "object") return null
  const o = payload as Record<string, unknown>
  if (typeof o.id === "string") return o.id
  return null
}

function readMessage(payload: unknown): string {
  if (!payload || typeof payload !== "object") return ""
  const o = payload as Record<string, unknown>
  if (typeof o.message === "string") return o.message
  return ""
}

File: backend-python/leadgen_backend/providers/hubspot.py
"""
HubSpot CRM provider — mirrors lib/providers/hubspot.ts

Supports contact upsert (search by email → PUT or POST) and note creation.
Mock-safe when HUBSPOT_API_KEY is not set.
"""
from __future__ import annotations

import os
from typing import Any

import httpx


_BASE = "https://api.hubapi.com"
_TIMEOUT = 12


def _is_configured() -> bool:
    return bool(os.getenv("HUBSPOT_API_KEY"))


def _headers() -> dict[str, str]:
    return {
        "Authorization": f"Bearer {os.getenv('HUBSPOT_API_KEY', '')}",
        "Content-Type": "application/json",
    }


def _contact_to_properties(prospect: dict[str, Any]) -> dict[str, str]:
    name = prospect.get("input_name") or ""
    parts = name.strip().split(None, 1)
    first = parts[0] if parts else ""
    last = parts[1] if len(parts) > 1 else ""
    return {
        "email": prospect.get("email") or "",
        "firstname": first,
        "lastname": last,
        "company": prospect.get("input_company") or "",
        "jobtitle": prospect.get("title") or "",
        "linkedin": prospect.get("input_linkedin_url") or "",
    }


async def push_hubspot_contact(prospect: dict[str, Any]) -> dict[str, Any]:
    """Upsert a contact in HubSpot by email. Returns {contact_id, created}."""
    if not _is_configured():
        mock_id = f"mock-{abs(hash(prospect.get('email', '')))}"
        return {"contact_id": mock_id, "created": False, "mock": True}

    email = (prospect.get("email") or "").strip()
    if not email:
        raise ValueError("Prospect has no email")

    # Search for existing contact
    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        search_resp = await client.post(
            f"{_BASE}/crm/v3/objects/contacts/search",
            headers=_headers(),
            json={
                "filterGroups": [{"filters": [{"propertyName": "email", "operator": "EQ", "value": email}]}],
                "properties": ["email"],
                "limit": 1,
            },
        )

    properties = _contact_to_properties(prospect)

    if search_resp.status_code == 200:
        results = search_resp.json().get("results", [])
        if results:
            contact_id = results[0]["id"]
            # Update existing
            async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
                await client.patch(
                    f"{_BASE}/crm/v3/objects/contacts/{contact_id}",
                    headers=_headers(),
                    json={"properties": properties},
                )
            return {"contact_id": contact_id, "created": False}

    # Create new
    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        create_resp = await client.post(
            f"{_BASE}/crm/v3/objects/contacts",
            headers=_headers(),
            json={"properties": properties},
        )

    if create_resp.status_code == 409:
        # Duplicate — fetch existing
        dup_id = create_resp.json().get("message", "").split("vid=")[-1].split(")")[0]
        return {"contact_id": dup_id or "unknown", "created": False}

    create_resp.raise_for_status()
    contact_id = create_resp.json().get("id", "unknown")
    return {"contact_id": contact_id, "created": True}


async def add_hubspot_note(contact_id: str, body: str) -> dict[str, Any]:
    """Add a note to a HubSpot contact."""
    if not _is_configured():
        return {"note_id": "mock-note", "mock": True}

    note_body = body[:10_000]  # HubSpot note body limit

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        resp = await client.post(
            f"{_BASE}/crm/v3/objects/notes",
            headers=_headers(),
            json={
                "properties": {"hs_note_body": note_body},
                "associations": [{
                    "to": {"id": contact_id},
                    "types": [{"associationCategory": "HUBSPOT_DEFINED", "associationTypeId": 202}],
                }],
            },
        )

    resp.raise_for_status()
    return {"note_id": resp.json().get("id", "unknown")}
